sympy/solvers/tests/test_solveset.py

```diff
@@ -2023,3 +2023,138 @@ def test_linear_coeffs():
+            (x, ImageSet(Lambda(n, 7*n + 4), S.Integers))
+    assert invert_modulo(Mod((x + 1)*(x + 2), 7), S(5), n, x) == \
+            ((x + 1)*(x + 2), ImageSet(Lambda(n, 7*n + 5), S.Integers))
+    # a.is_Pow
+    assert invert_modulo(Mod(x**4, 7), S(5), n, x) == \
+            (x**4, ImageSet(Lambda(n, 7*n + 5), S.Integers))
+    assert invert_modulo(Mod(3**x, 4), S(3), n, x) == \
+            (x, ImageSet(Lambda(n, 2*n + 1), S.Naturals0))
+    assert invert_modulo(Mod(2**(x**2 + x + 1), 7), S(2), n, x) == \
+            (x**2 + x + 1, ImageSet(Lambda(n, 3*n + 1), S.Naturals0))
+
+
+def test_solve_modular():
+    x = Symbol('x')
+    n = Dummy('n')
+    # if rhs has symbol (need to be implemented in future).
+    assert solveset(Mod(x, 4) - x, x) == \
+            ConditionSet(x, Eq(-x + Mod(x, 4), 0), \
+            S.Complexes)
+    # when _invert_modular fails to invert
+    assert solveset(3 - Mod(sin(x), 7), x) == \
+            ConditionSet(x, Eq(Mod(sin(x), 7) - 3, 0), S.Complexes)
+    assert solveset(3 - Mod(log(x), 7), x) == \
+            ConditionSet(x, Eq(Mod(log(x), 7) - 3, 0), S.Complexes)
+    assert solveset(3 - Mod(exp(x), 7), x) == \
+            ConditionSet(x, Eq(Mod(exp(x), 7) - 3, 0), S.Complexes)
+    # EmptySet solution definitely
+    assert solveset(7 - Mod(x, 5), x) == EmptySet()
+    assert solveset(5 - Mod(x, 5), x) == EmptySet()
+    # Negative m
+    assert solveset(2 + Mod(x, -3), x) == \
+            Intersection(S.Complexes, ImageSet(Lambda(n, -3*n - 2), S.Integers))
+    assert solveset(4 + Mod(x, -3), x) == EmptySet()
+    # linear expression in Mod
+    assert solveset(3 - Mod(x, 5), x) == ImageSet(Lambda(n, 5*n + 3), S.Integers)
+    assert solveset(3 - Mod(5*x - 8, 7), x) == \
+                ImageSet(Lambda(n, 7*n + 5), S.Integers)
+    assert solveset(3 - Mod(5*x, 7), x) == \
+                ImageSet(Lambda(n, 7*n + 2), S.Integers)
+    # higher degree expression in Mod
+    assert solveset(3 - Mod(x**2 - 2*x + 1, 7), x) == \
+            Union(Intersection(S.Complexes, ImageSet(Lambda(n, \
+            sqrt(7*n + 3) + 1), S.Integers)), Intersection(S.Complexes, \
+            ImageSet(Lambda(n, 1 - sqrt(7*n + 3)), S.Integers)))
+    assert solveset(3 - Mod(3*x**2 + 5*x + 1, 7), x) == \
+            Union(Intersection(S.Complexes, ImageSet(Lambda(n, \
+            -sqrt(7)*sqrt(12*n + 7)/6 - S(5)/6), S.Integers)), \
+            Intersection(S.Complexes, ImageSet(Lambda(n, \
+            sqrt(7)*sqrt(12*n + 7)/6 - S(5)/6), S.Integers)))
+    assert solveset(3 - Mod(x**4, 7), x) == \
+            Union(Intersection(S.Complexes, ImageSet(Lambda(n, \
+            (7*n + 3)**(S(1)/4)), S.Integers)), Intersection(S.Complexes, \
+            ImageSet(Lambda(n, I*(7*n + 3)**(S(1)/4)), S.Integers)), \
+            Intersection(S.Complexes, ImageSet(Lambda(n, -(7*n + 3)**(S(1)/4)), \
+            S.Integers)), Intersection(S.Complexes, ImageSet(Lambda(n, \
```

Comment:
Sorry for keeping you busy. Could you point me to some theoretical references for the above types of equations?

Let's say I'm new to this and discover that SymPy should be able to solve modular equations. My first thought will be number theory / algebra, so I might try this:
```
In [33]: solveset(Mod(x**2, 4), x, S.Integers)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/root/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    192         eq = p.as_expr()
--> 193         assert eq.is_polynomial()
    194     except (GeneratorsNeeded, AssertionError, AttributeError):

AssertionError:

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-33-a3d26480b232> in <module>()
----> 1 solveset(Mod(x**2, 4), x, S.Integers)

/root/sympy/sympy/solvers/solveset.py in solveset(f, symbol, domain)
   1939             try:
   1940                 r = Dummy('r', **assumptions)
-> 1941                 return solveset(f.xreplace({symbol: r}), r, domain
   1942                     ).xreplace({r: symbol})
   1943             except InconsistentAssumptions:
[... snip...]
/root/sympy/sympy/sets/handlers/intersection.py in intersection_sets(self, other)
    257             a, b = Dummy('a'), Dummy('b')
    258             fa, ga = f.subs(n, a), g.subs(m, b)
--> 259             solns = list(diophantine(fa - ga))
    260             if not solns:
    261                 return EmptySet()

/root/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    194     except (GeneratorsNeeded, AssertionError, AttributeError):
    195         raise TypeError(filldedent('''
--> 196     Equation should be a polynomial with Rational coefficients.'''))
    197
    198     # permute only sign

TypeError:
Equation should be a polynomial with Rational coefficients.
```
Then I might try to leave the `S.Integers` aside, only to find that the power is now interpreted as a function over the reals/complexes:
```
In [34]: solveset(Mod(x**2, 4), x)
Out[34]: ({-2⋅√n | n ∊ ℤ} ∩ ℂ) ∪ ({2⋅√n | n ∊ ℤ} ∩ ℂ)
```
which, in a way, is not outright wrong, but it's not helpful either. Is it useful at all to accept non-integers? (This is an honest question, I'm trying to understand what this PR accomplishes and what not.)

Compare it to this:
```
In [45]: _solve_modular(Mod(3*x + 1, 5), x, S.Reals)
Out[45]: {5⋅n + 3 | n ∊ ℤ}
```
If all reals are potential solutions, why doesn't the solution set include `4/3` or `14/3`? (This is obviously a rhetorical question. I'm not arguing in favor of having them added.)

What are the precise types of mathematical equations you're targeting?

Reply:
First of all this error is arising because of ImageSet Intersection for which I have raised the issue already(thats why I had not added any test with domain Integers).

> Could you point me to some theoretical references for the above types of equations?

I had not any theoretical references. I check them by substituting or from wolfram alpha.
The algorithm I am implementing is based on heurestics and it will take several iterations to improve it.

> What are the precise types of mathematical equations you're targeting?

As mentioned in documentation of functions.
```
A - Mod(B, C) = 0

    A -> This can or cannot be a function of symbol.
    B -> This is surely a function of symbol.
    C -> It is an integer.
```
Also the same thing get struck in my mind when I you draw my attention toward S.Complexes domain. May be I think I should only let equations enter which have domain Integers. What do you think because I also see no use when domain is S.Reals and S.Complexes.

Reply:
> First of all this error is arising because of ImageSet Intersection for which I have raised the issue already(thats why I had not added any test with domain Integers).

Oh, right. So that issue is really disrupting quite a part of your project? That must be painful. I'm sorry.

> ```
> A - Mod(B, C) = 0
> 
>     A -> This can or cannot be a function of symbol.(But currently its not a function of x)
>     B -> This is surely a function of symbol.
>     C -> It is an integer.
> ```

Well, "function" is not very precise... Powers? (which exponents?) Polynomials? (which degrees?) Transcendental?

> Also the same thing get struck in my mind when you draw my attention toward S.Complexes domain. May be I think I should only let equations enter which have domain Integers. What do you think because I also see no use when domain is S.Reals and S.Complexes.

In my opinion that would be best. I really hope you're not feeling that I'm "sabotaging" your GSOC project, I checked this PR out simply out of interest.

If you're restricting the PR to integer relations, the good news is:
- the problematic intersection could go away (AFAICT it will be superfluous);
- you could probably make good use of some functions from the `ntheory` module, notably `sqrt_mod` and `nthroot_mod`.

Reply:
> So that issue is really disrupting quite a part of your project?

Yeah it is affecting very much.
> I really hope you're not feeling that I'm "sabotaging" your GSOC project, I checked this PR out simply out of interest.

No need to feel that way. After all we all want best for SymPy irrespective of GSoC. So I would encourage you to also suggest some good ideas or equations which we can cover which might not strike in my mind and moreover we can also do discussion of mod in S.Complexes also.

New changes will be accomodated in next commits tomorrow.

Reply:
> So I would encourage you to also suggest some good ideas or equations which we can cover

IMHO you should get the basics right first, that is returning meaningful answers for the following:
```
solveset(Modular(x**2, 160) - 9, x, Integers)
solveset(Modular(x**4, 17) - 13, x, Integers)
solveset(Modular(7**x, 41) - 15, x, Integers)
```
(see `sqrt_mod`, `nthroot_mod`, `discrete_log`)

For the first one, here are some pointers:
```
In [15]: sqrt_mod(9, 160, all_roots=True)
Out[15]: [3, 13, 67, 77, 83, 93, 147, 157]

In [16]: s = FiniteSet(*_)

In [17]: ConditionSet(n, Contains(Mod(n, 160), s))
Out[17]: {n | (n mod 160 ∈ {3, 13, 67, 77, 83, 93, 147, 157})}
```
(I think ImageSets or unions thereof will quickly become too unwieldy as solution sets. Not sure.)

Beware: see notes in ntheory.py; check ntheory for possible exceptions; bugs (#17373)
Also: add lots of tests (that you carefully checked), include various moduli (prime, composite, etc.)